File: python3/find_first_and_last_position_of_element_in_sorted_array.py
from typing import List

"""
Space   : O(1)
Time    : O(log n)
"""

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        ans = [-1, -1]
        if not nums:
            return [-1, -1]

        n = len(nums)
        left, right = 0, n-1

        while left < right:
            mid = (left + right) // 2
            if nums[mid] < target:
                left = mid + 1
            else:
                right = mid

        if nums[right] != target:
            return [-1, -1]

        ans[0] = right

        right = n-1
        while left < right:
            mid = ((left + right) // 2) + 1
            if nums[mid] <= target:
                left = mid
            else:
                right = mid - 1

        ans[1] = right
        return ans

        

File: golang/find_first_and_last_position_of_element_in_sorted_array.go
package main

// # Space   : O(1)
// # Time    : O(log n)

func searchRange(nums []int, target int) []int {
	ans := []int{-1, -1}
	if len(nums) == 0 {
		return ans
	}

	s, m, e := 0, 0, len(nums)-1
	for s < e {
		m = (s + e) / 2
		if nums[m] < target {
			s = m + 1
		} else {
			e = m
		}
	}

	if nums[e] != target {
		return ans
	}
	ans[0] = e

	e = len(nums) - 1
	for s < e {
		m = ((s + e) / 2) + 1
		if nums[m] <= target {
			s = m
		} else {
			e = m - 1
		}
	}

	ans[1] = e
	return ans
}
